PYIC-5868: Modify how build-proven-user-identity-details in core-back sends user data

sammyiyke opened the pull request:
<!-- Provide a general summary of your changes in the Title above -->
<!-- Include the Jira ticket number in square brackets as prefix, eg `[P4-XXXX] PR Title` -->

## Proposed changes
[PYIC-5868](https://govukverify.atlassian.net/browse/PYIC-5868) - Modify how build-proven-user-identity-details in core-back sends user data

### What changed

Adds formatted name to the data returned by the `BuildProvenUserIdentityDetailsHandler` lambda.

This PR add new field `formattedName` to the data returned by  `BuildProvenUserIdentityDetailsHandler`. This field contains the name and name parts (GivenName and LastName).

<!-- Describe the changes in detail - the "what"-->

### Why did it change

We need to send the formatted name including name parts to the core-front when a user details is requested in `BuildProvenUserIdentityDetailsHandler`

This is required to support the new COI journey for confirming and changing user details

This implementation avoids changing or manipulating (e.g regex or splicing in front-end) the full name sent to core-front but adds a new field that contains the name broken into the different name parts. Avoiding potential complications with name splitting in core-front

<!-- Describe the reason these changes were made - the "why" -->

### Issue tracking
<!-- List any related Jira tickets or GitHub issues -->
<!-- List any related ADRs or RFCs -->
<!-- Delete/copy as appropriate -->

- [PYIC-5868](https://govukverify.atlassian.net/browse/PYIC-5868)
- [PYIC-5669](https://govukverify.atlassian.net/browse/PYIC-5669)

## Checklists

### Environment variables or secrets

<!-- Delete if changes DO include new environment variables or secrets -->
- [x] No environment variables or secrets were added or changed



[PYIC-5868]: https://govukverify.atlassian.net/browse/PYIC-5868?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[PYIC-5868]: https://govukverify.atlassian.net/browse/PYIC-5868?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

sammyiyke committed:
PYIC-5868 - Adds formatted name to `BuildProvenUserIdentityDetailsHandler`

This adds formatted name to the data returned by the `BuildProvenUserIdentityDetailsHandler` lambda.

The formatted name is returned as a map of the name parts and values

This implementation retains support for full name value returned in name as used in core-front but adds new `formattedName` to the data returned

sammyiyke committed:
PYIC-5868 - Tests for new `formattedName` in `BuildProvenUserIdentityDetailsHandler` and type `Name`

This adds tests for new field `formattedName` in the data returned by the `BuildProvenUserIdentityDetailsHandler` lambda.

Adds tests for new method `getFormattedName`

sammyiyke committed:
PYIC-5868 - ran spotlessApply

Joe-Edwards-GDS commented on libs/common-services/src/main/java/uk/gov/di/ipv/core/library/domain/Name.java:
```diff
@@ -33,4 +35,13 @@ public String getFullName() {
+        for (NameParts namePart : nameParts) {
+            formattedName.put(namePart.getType(), namePart.getValue());
+        }
+        return formattedName;
```
I don't think this will work - you can have multiple name parts of the same type. Have a look at the examples linked on https://vocab.account.gov.uk/v1/classes/NameClass/ , e.g.:
```
{
  "nameParts": [
    {
      "value": "Ana",
      "type": "GivenName"
    },
    {
      "value": "Maria",
      "type": "GivenName"
    },
    {
      "value": "Carvalho",
      "type": "FamilyName"
    },
    {
      "value": "Sousa",
      "type": "FamilyName"
    }
  ]  
}
```

Joe-Edwards-GDS replied:
TBH I would probably suggest that you just pass `nameParts` directly to the frontend, and the it can handle things as it sees fit.

Joe-Edwards-GDS commented on lambdas/build-proven-user-identity-details/src/main/java/uk/gov/di/ipv/core/buildprovenuseridentitydetails/domain/ProvenUserIdentityDetails.java:
```diff
@@ -9,22 +9,26 @@
 @Builder
 public class ProvenUserIdentityDetails {
     @JsonProperty private final String name;
+    @JsonProperty private final Map<String, String> formattedName;
```
I'm not sure about `formattedName` here - I think the original `name` is actually the formatted one, and I'm also not sure about having it just be a map.

As below, I wonder about just making this `List<NamePart> nameParts`?

sammyiyke committed:
PYIC-5868 - PR Related changes

- Renamed formattedName to nameParts
- Changed related methods to match new format
- Removes `getFormattedName` method in `Name` class
- Sends `nameParts` as is - `NameParts` - to front-end
- Fixes tests

sammyiyke committed:
PYIC-5868 - Adds test

- Adds test for getNameParts

Joe-Edwards-GDS commented on lambdas/build-proven-user-identity-details/src/main/java/uk/gov/di/ipv/core/buildprovenuseridentitydetails/domain/NameAndDateOfBirth.java:
If you like, simple classes like this can be turned into record classes, but it's fine to leave:

```
public record NameAndDateOfBirth(String name, List<NameParts> nameParts, String dateOfBirth) {}
```

sammyiyke replied:
Sounds interesting. Easier code to read too. 

Might take some time to explore that option

sammyiyke replied:
Implemented this. It's cleaner and less code 😃

Joe-Edwards-GDS reviewed (dismissed):
Thanks, I think this should be more flexible :)

sammyiyke committed:
PYIC-5868 - PR Related changes

Turned `NameAndDateOfBirth` class into a record class
